### PR TITLE
Add R_ARM_CALL runnable test

### LINKED ISSUE #1224: [ARM] Add static runnable test for relocations
[ARM] Add static runnable test for relocations

### PR DESCRIPTION
**Pull request by Steven6798**
Fix qualcomm#1224

### CODE REVIEW

**Comment by Steven6798** [comment]
I'll appreciate some feedback on this, as this will serve as my template for adding more relocation tests. I'm sure these changes can be improved even more.

File: test/Common/RunTests/MyTest/MyTest.test [hunk 1]
```diff
@@ -0,0 +1,29 @@
+REQUIRES: run_test
```
Comment (quic-areg) [inline test/Common/RunTests/MyTest/MyTest.test]: `MyTest` is a placeholder name; it should be changed to something descriptive - probably the name of the relocation in this case.

This is an ARM-specific test; I don't think it should belong in a directory called `Common`.

File: test/ARM/RunTests/CallReloc/CallReloc.test [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+# Test the ARM R_ARM_CALL relocation.
+#END_COMMENT
+#START_TEST
+#RUN: %run_cc -o %t1.o -c %p/Inputs/1.s
+# Partial
+#RUN: %run_cc -o %t1.o %t1.o -r
+#RUN: llvm-readelf -r %t1.o | %filecheck %s -check-prefix RELOCS
+# Static
+#RUN: %run_cc -o %t1.out %t1.o -static
+#RUN: llvm-objdump -d --no-show-raw-insn %t1.out | FileCheck %s -check-prefix STATIC_RELOCS
```
Comment (quic-seaswara) [inline test/ARM/RunTests/CallReloc/CallReloc.test]: %objdump

File: test/ARM/RunTests/CallReloc/Inputs/1.s [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+                                        @ -- End function
+	.type	.L.str,%object                  @ @.str
+	.section	.rodata.str1.1,"aMS",%progbits,1
+.L.str:
+	.asciz	"Hello World"
+	.size	.L.str, 12
+
+	.ident	"clang version 23.0.0git (https://github.com/llvm/llvm-project.git 832285ab7dffc8381ae76182b9b1f0f4cbae23d0)"
+	.section	".note.GNU-stack","",%progbits
+	.addrsig
```
Comment (quic-seaswara) [inline test/ARM/RunTests/CallReloc/Inputs/1.s]: prefer 'C' code instead of assembly.

File: docs/DeveloperDocs/ARM/relocations.md [hunk 4]
```diff
@@ -0,0 +1,15 @@
+# ARM Relocations
+
+## Relocation Table
+
+| Code | Name       | Type   | Class | Operation           |
+|------|------------|--------|-------|---------------------|
+| 28   | R_ARM_CALL | Static | Arm   | ((S + A) \| T) - P  |
```
Comment (quic-seaswara) [inline docs/DeveloperDocs/ARM/relocations.md]: add a link to the ARM ELF ABI. 

Can we have the table to also show the number of bits available and when the linker decides to insert a trampoline.

Also it may be useful to consider the differences between

* Static Linking
* PIE
* Shared objects

to make the table complete.

**Review by quic-seaswara — CHANGES_REQUESTED** [review]

File: test/Common/RunTests/MyTest/MyTest.test [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+#---MyTest.test--------------------- Executable------------------#
+#BEGIN_COMMENT
+# Test the ARM R_ARM_CALL relocation.
+#END_COMMENT
+#START_TEST
+#RUN: %run_cc -o %t1.o -c %p/Inputs/1.s
+# Partial
+#RUN: %run_cc -o %t1.o %t1.o -r
+#RUN: llvm-readelf -r %t1.o | %filecheck %s -check-prefix RELOCS
```
Comment (quic-areg) [inline test/Common/RunTests/MyTest/MyTest.test]: The `readelf` and `objdump` checks belong in traditional static lit tests rather than runtime tests. Runtime tests should only check runtime behavior and the two test types should not overlap, imo.

File: test/ARM/RunTests/CallReloc/Inputs/1.s (hunk 3 above)
Comment (Steven6798) [reply test/ARM/RunTests/CallReloc/Inputs/1.s]: Sounds good. I'll use C code as much as I can and assembly only when needed.

File: test/ARM/RunTests/CallReloc/CallReloc.test (hunk 2 above)
Comment (Steven6798) [reply test/ARM/RunTests/CallReloc/CallReloc.test]: Thanks!

File: test/Common/RunTests/MyTest/MyTest.test [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+#RUN: %run %t1.out | %filecheck %s
+# Dynamic
+#RUN: %run_cc -o %t1.out %t1.o -Wl,-dy
+#RUN: llvm-objdump -d --no-show-raw-insn %t1.out | FileCheck %s -check-prefix DYN_RELOCS
+#RUN: %run %t1.out | %filecheck %s
+
+#RELOCS: Relocation section '.rel.text'
+#RELOCS: {{.*}} R_ARM_CALL
+#STATIC_RELOCS: bl	0x8049158 <puts>
+#DYN_RELOCS: bl	0x3a4 <__cxa_finalize@plt+0xc>
```
Comment (quic-areg) [inline test/Common/RunTests/MyTest/MyTest.test]: These hard-coded address checks may be brittle (but I don't think they belong in runnable tests)

File: test/Common/RunTests/MyTest/MyTest.test [hunk 7]
```diff
@@ -0,0 +1,29 @@
+REQUIRES: run_test
+#---MyTest.test--------------------- Executable------------------#
+#BEGIN_COMMENT
+# Test the ARM R_ARM_CALL relocation.
+#END_COMMENT
+#START_TEST
+#RUN: %run_cc -o %t1.o -c %p/Inputs/1.s
+# Partial
```
Comment (quic-areg) [inline test/Common/RunTests/MyTest/MyTest.test]: I don't see what the partial link step achieves here.

File: test/Common/RunTests/MyTest/Inputs/1.s [hunk 8]
```diff
@@ -0,0 +1,54 @@
+	.syntax	unified
```
Comment (quic-areg) [inline test/Common/RunTests/MyTest/Inputs/1.s]: This input might be more readable as 'C'. If assembly is kept, it can be greatly simplified. Test inputs, whether C or asm, should be minimal. Eg these sections are not required:

```
.ident	"clang version 23.0.0git (https://github.com/llvm/llvm-project.git 832285ab7dffc8381ae76182b9b1f0f4cbae23d0)"	
.section	".note.GNU-stack","",%progbits	.addrsig
```

File: docs/DeveloperDocs/ARM/relocations.md [hunk 9]
```diff
@@ -0,0 +1,15 @@
+# ARM Relocations
```
Comment (quic-areg) [inline docs/DeveloperDocs/ARM/relocations.md]: I do not see the point of the docs. We are duplicating the ARM ABI which is already well-documented [here](https://github.com/ARM-software/abi-aa/blob/main/aaelf32/aaelf32.rst). I don't think our documentation will be any better than the official docs, nor will it be maintained as well. 

This doc so far contains general ARM facts, not eld specifics. Official ARM sources are probably a better source for this information.

**Review by quic-areg — COMMENTED** [review]
In general, I think we should prefer C inputs for runnable tests, get rid of the docs if they are going to just parrot the ABI, and separate the types of checks we make in static vs runtime tests.

File: test/Common/RunTests/MyTest/MyTest.test (hunk 1 above)
Comment (Steven6798) [reply test/Common/RunTests/MyTest/MyTest.test]: Yes, that was a placeholder for my initial testing. I forgot to update that. It's fixed. Thanks!

File: test/Common/RunTests/MyTest/MyTest.test (hunk 5 above)
Comment (Steven6798) [reply test/Common/RunTests/MyTest/MyTest.test]: Got it. Thanks!

File: test/Common/RunTests/MyTest/MyTest.test (hunk 6 above)
Comment (Steven6798) [reply test/Common/RunTests/MyTest/MyTest.test]: They were indeed brittle but will be removed as you said. Thanks!

File: test/Common/RunTests/MyTest/Inputs/1.s (hunk 8 above)
Comment (Steven6798) [reply test/Common/RunTests/MyTest/Inputs/1.s]: Thanks. I'll use C as much as I can and only use assembly when needed. In those corner cases I'll clean the assembly file as much as possible.

File: docs/DeveloperDocs/ARM/relocations.md (hunk 9 above)
Comment (Steven6798) [reply docs/DeveloperDocs/ARM/relocations.md]: I agree. @quic-seaswara thoughts?

Comment (quic-seaswara) [reply docs/DeveloperDocs/ARM/relocations.md]: I am thinking that to use a single point of reference to show what relocations are implemented by eld, there are a lot of relocations, that we have not seen. This is useful for code review and any changes that we need to make to support bug fixes. 

We should definitely look to add the eld specifics to the relocation, number of bits, does it create a PLT or no PLT for shared objects and PLT/no PLT for PIE's etc.

**Comment by quic-seaswara** [comment]
Rename the test to R_ARM_CALL.test

**Comment by Steven6798** [comment]
> Rename the test to R_ARM_CALL.test

Sounds good. Thanks!

File: docs/DeveloperDocs/ARM/relocations.md (hunk 9 above)
Comment (Steven6798) [reply docs/DeveloperDocs/ARM/relocations.md]: That make sense. I'll only add ELD specific information and remove any generic information documented in the ARM ABI.

File: test/Common/RunTests/MyTest/MyTest.test (hunk 7 above)
Comment (Steven6798) [reply test/Common/RunTests/MyTest/MyTest.test]: Just testing that the relocation was still there after partial link. But as you said, these test don't belong in runnable tests. Thanks!

Comment (quic-seaswara) [reply test/Common/RunTests/MyTest/MyTest.test]: taking the output of partial links and building an executable / shared object and running it belongs, if that is what you were looking to do.

Comment (Steven6798) [reply test/Common/RunTests/MyTest/MyTest.test]: Got it. Then I'll keep the partial link and remove the additional checks so we can still test partial linking. Thanks!

Comment (quic-seaswara) [reply test/Common/RunTests/MyTest/MyTest.test]: you need the partial linked object file to be used in a shared library / executable link and run it. This will ensure that eld did not break some assumption in partial link.

File: docs/DeveloperDocs/ARM/relocations.md (hunk 9 above)
Comment (Steven6798) [reply docs/DeveloperDocs/ARM/relocations.md]: So far, I don't see anything eld specific for R_ARM_CALL. It looks like it is fully supported. Thoughts?

**Comment by Steven6798** [comment]
I addressed the issues. Let me know if I missed something or if I can improve it even more.

**Review by quic-areg — COMMENTED** [review]
LGTM mostly.

For future PRs, if there is no existing static test for the relocation it might be a good idea to add one in addition to the runtime test. `R_ARM_CALL` is already well tested, so that does not apply here, but it might for future relocations.

The doc looks better than before imo, but I still think it adds little value. Checking the list of supported relocations for a reviewer or developer was already a pretty trivial task (they are all defined in *RelocationFunctions.h) , and I worry that the doc risks going stale.

File: test/ARM/RunTests/R_ARM_CALL/R_ARM_CALL.test [hunk 10]
```diff
@@ -0,0 +1,21 @@
+REQUIRES: run_test
+#---R_ARM_CALL.test--------------------- Executable------------------#
+#BEGIN_COMMENT
+# Test the ARM R_ARM_CALL relocation.
+#END_COMMENT
+#START_TEST
+#RUN: %run_cc -o %t1.o -c %p/Inputs/1.c
```
Comment (quic-seaswara) [inline test/ARM/RunTests/R_ARM_CALL/R_ARM_CALL.test]: can we establish that this generates a R_ARM_CALL relocation ?

Comment (quic-seaswara) [inline test/ARM/RunTests/R_ARM_CALL/R_ARM_CALL.test]: compile with -fPIC and -fpic options

File: test/ARM/RunTests/R_ARM_CALL/R_ARM_CALL.test [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+# Partial
+#RUN: %run_cc -o %t1.o %t1.o -r
+# Static
+#RUN: %run_cc -o %t1.out %t1.o -static
+#RUN: %run %t1.out | %filecheck %s
```
Comment (quic-seaswara) [inline test/ARM/RunTests/R_ARM_CALL/R_ARM_CALL.test]: This brings in all the libraries etc along with testing for a reloc, is this designed as intended ? 

The sysroot might bring in tons of relocations such as R_ARM_CALL already.

File: docs/DeveloperDocs/ARM/relocations.md [hunk 12]
```diff
@@ -0,0 +1,9 @@
...
+
+Information about ARM relocations can be found here: https://github.com/ARM-software/abi-aa/blob/main/aaelf32/aaelf32.rst
+
+## Status
+
+| Name       | Status          |
+|------------|-----------------|
+| R_ARM_CALL | Fully supported |
```
Comment (quic-seaswara) [inline docs/DeveloperDocs/ARM/relocations.md]: We need eld specifics here. 

Total bits available in the relocation, when does the linker create trampolines for branch range overflows, does it create a PLT for shared library links, how does it work for local symbols or non premptible symbols.

**Review by quic-seaswara — CHANGES_REQUESTED** [review]

File: test/ARM/RunTests/R_ARM_CALL/R_ARM_CALL.test (hunk 10 above)
Comment (Steven6798) [reply test/ARM/RunTests/R_ARM_CALL/R_ARM_CALL.test]: I can add the llvm-readelf again to check for the relocation like I was doing before. Is that enough?

Comment (Steven6798) [reply test/ARM/RunTests/R_ARM_CALL/R_ARM_CALL.test]: Sounds good

File: test/ARM/RunTests/R_ARM_CALL/R_ARM_CALL.test (hunk 11 above)
Comment (Steven6798) [reply test/ARM/RunTests/R_ARM_CALL/R_ARM_CALL.test]: In theory we would like to isolate this test as much as possible. The c code uses libraries, so we need to include them and also test static linking. Is there a better way to do this without including so much code?

Comment (quic-seaswara) [reply test/ARM/RunTests/R_ARM_CALL/R_ARM_CALL.test]: https://dev.to/ripan030/my-first-bare-metal-program-from-reset-to-hello-4cn8 is an example

**Comment by Steven6798** [comment]
I'm making progress. I'll leave the documentation for another issue to focus on 1 thing at a time. I improved the testing coverage of this relocation. I also started using Bare-Metal runs as much as possible to isolate the tests.

**Comment by Steven6798** [comment]
I also updated the lit.cfg to include qemu-system-target to use Bare-Metal runs

File: .github/workflows/ci.yml [hunk 13]
```diff
@@ -129,6 +129,12 @@ jobs:
         working-directory: pr-${{ env.PR_NUMBER }}/obj
         run: ccache --evict-older-than 7200s --dir=$CCACHE_DIR
 
+      - name: Install dependencies
+        shell: bash
+        run:  |
+          sudo apt update
+          sudo apt install -y libsdl2-2.0-0
```
Comment (parth-07) [inline .github/workflows/ci.yml:136]: Why do we need this?

File: test/lit.cfg [hunk 14]
```diff
@@ -214,10 +214,12 @@ dirname = 'dirname'
 datalayout = ''
 link_path = which(link)
 run=""
+run_sys=""
 run_cc = ''
 run_cxx = ''
 sysroot = ''
 emulator = ''
+emulator_system = ''
```
Comment (parth-07) [inline test/lit.cfg:222]: Can you please explain why do we need system emulator? For which cases linux emulator is not enough?

File: test/ARM/RunTests/R_ARM_CALL/Inputs/hidden_callee.c [hunk 15]
```diff
@@ -0,0 +1,22 @@
+// Hidden symbol: non-preemptible, direct relocation, no PLT entry.
+#define UART0_DR ((volatile unsigned int *)0x10009000)
```
Comment (parth-07) [inline test/ARM/RunTests/R_ARM_CALL/Inputs/hidden_callee.c:2]: Can we use printf and use non-system qemu emulator?

File: .github/workflows/ci.yml (hunk 13 above)
Comment (Steven6798) [reply .github/workflows/ci.yml:136]: I'm trying to use the qemu system emulator. the machine that runs the CI workflow is missing a library. Shakti suggested adding it to the workflow but clearly that did not work. If you have any suggestions about how to address this let me know.

File: test/lit.cfg (hunk 14 above)
Comment (Steven6798) [reply test/lit.cfg:222]: Shankar suggested using the system emulator to isolate the testing as much as possible. For example, when static linking a lot gets included in the binary, with this approach, only a minimal amount of code is included. What do you think?

File: test/ARM/RunTests/R_ARM_CALL/Inputs/hidden_callee.c (hunk 15 above)
Comment (Steven6798) [reply test/ARM/RunTests/R_ARM_CALL/Inputs/hidden_callee.c:2]: We are discussing using qemu system in another comment so I'll close this one.

File: test/lit.cfg (hunk 14 above)
Comment (parth-07) [reply test/lit.cfg:222]: @quic-seaswara Hi Shankar, I am concerned that writing baremetal code and using system emulator requires more boilerplate and makes testing more difficult without providing any apparent benefit. 

> to isolate the testing as much as possible

We are only removing the libc dependency -- all other dependency are still there, and we are unlikely to encounter bugs/corner-cases in libc when doing extended testing of different relocations computations because we won't be using any advanced feature of libc.

> For example, when static linking a lot gets included in the binary, with this approach, only a minimal amount of code is included. 

I agree with this, but I am not sure if less size is worth the benefit here. We can always delete the test artifacts after the testing.

**Comment by Steven6798** [comment]
The way the ARM testing is going to be done changed, I will close this PR and open a new one.
